modules/compute/gke-node-pool/main.tf

```diff
@@ -150,6 +150,12 @@ resource "google_container_node_pool" "node_pool" {
         "net.ipv4.tcp_wmem" = "4096 16384 16777216"
       }
     }
+
+    reservation_affinity {
+      consume_reservation_type = var.reservation_type
```

Comment:
Does this allow customer to not specify reservation affinity at all ?

Reply:
Thanks for the comment. PR in its current state will not. 

Still working on the defaults considering possible scenarios. Specifically, treatment of null vs empty strings by the provider. (being my fist Terraform PR) going to rely on tests to come up with those.

For `consume_reservation_type`, `NO_RESERVATION` is a good default. It will take care of no reservation affinity at all (per the provider documentation) as long as `key`, `values` defaults are such that provider ignores them.

Reply:
I think there is this pattern here : https://github.com/GoogleCloudPlatform/cluster-toolkit/blob/9e68ecc363ebf212c3f574db9b25aa2f86f6db97/community/modules/scheduler/schedmd-slurm-gcp-v6-controller/controller.tf#L141

```
count = var.cloudsql != null ? 1 : 0
```

Similar to this : https://github.com/hashicorp/terraform/issues/21953

Can you see if that helps ?

Reply:
This pattern is not applicable here. It is used to disable a "resource". What we are trying to do is to conditionally not include an "argument".

Arguments with `null` values are treated unset and omitted by Terraform ([documentation reference](https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements#conditionally-omitted-arguments:~:text=Terraform%20v0.12%20now%20allows%20assigning%20the%20special%20value%20null%20to%20an%20argument%20to%20mark%20it%20as%20%22unset%22.)). Checked this with by creating a terraform plan. Kept `key`, `values` null with `NO_RESERVATION` as the default value of `consume_reservation_type`. In the generated plan, only the `consume_reservation_type` argument is part of the `reservation_affinity` block

Reply:
Got, does this pattern work ? 

```
dynamic "gcfs_config" {
      for_each = var.enable_gcfs ? [1] : []
      content {
        enabled = true
      }
```

or https://github.com/GoogleCloudPlatform/cluster-toolkit/blob/e01aa0d2de300ed3e9f7ec3eb484a5fe06546da0/modules/compute/gke-node-pool/main.tf#L131

I am also a bit new to terraform so if the above pattern also doesn't work, let's ask around. (Hopefully it is a not a huge change)

Reply:
> This pattern is not applicable here. It is used to disable a "resource". What we are trying to do is to conditionally not include an "argument".
> 
> Arguments with `null` values are treated unset and omitted by Terraform ([documentation reference](https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements#conditionally-omitted-arguments:~:text=Terraform%20v0.12%20now%20allows%20assigning%20the%20special%20value%20null%20to%20an%20argument%20to%20mark%20it%20as%20%22unset%22.)). Checked this with by creating a terraform plan. Kept `key`, `values` null with `NO_RESERVATION` as the default value of `consume_reservation_type`. In the generated plan, only the `consume_reservation_type` argument is part of the `reservation_affinity` block

I meant, the `null` arguments should work for us.